feat: add function workflow to tutorial

lance opened the pull request:
This commit adds a functions section to the tutorial. It walks the user through creating and deploying a function for HTTP. Then it walks through some minor integrations with Knative Eventing with a CloudEvent function. Finally, installation instructions for `func` and `kn func` are added, as well as a top-level "Functions" menu item that has a (still needs to be improved) section on some function next steps.

Fixes: https://github.com/knative/docs/issues/5205
Fixes: https://github.com/knative/docs/issues/5265
Fixes: https://github.com/knative-sandbox/kn-plugin-func/issues/903

Signed-off-by: Lance Ball <lball@redhat.com>

csantanapr commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,26 @@
+    Create the function by running the command:
+
+    ```bash
+    func create -l go -t cloudevents hello
````
I would use `kn func` instead of func

csantanapr replied:
The installation of func plugin for `kn` are missing in the tutorial, maybe they should be after install kn, go ahead and install the func plugin.

lance replied:
> I would use `kn func` instead of func

I am curious why. To me that seems to add additional friction. The user is required to install `kn`, then install `func`, and make sure that `func` is installed in the right way/place in order for it to work as a plugin correctly.

abrennan89 replied:
+1 for `kn func`, mostly because it's just more consistent from a UX perspective. Adding `func` as another CLI in its own right rather than it being a plugin of `kn` to me is more confusing and will lead to more questions, e.g. is `kn` required at all then or should users just use `func` and YAML / `kubectl`? I think from a user perspective, even though there might be an extra step in the setup, `kn func` is easier to remember.

lance replied:
I just don't think it's nearly as easy to set up `kn func` as it is to simply use `func`. To set up `kn func` the user has two downloads, then must place each of them in the correct location on their system. Additionally, it's not always clear where `kn` picks up the plugins from. Let's say a user has `func` in both `~/bin/` and `/opt/local/bin` and they are two different versions. Which one will `kn` use? It's not obvious. Whereas, with simply `func` the user has to type `brew install func` and it's done - or download only `func` from an upstream release - in either case, there is much less opportunity for user error and confusing messages.

salaboy replied:
I would use `func` too.. adding `kn` and treating `func` just as a plugin feels unnatural to me, if you are only working with functions.

lkingland replied:
The simplest path is to stick with `func` here, and call out to a plugin installation guide as necessary.

I understand the desire to explain and introduce everything we have, from the bottom up, but every extra command that we expect a user to run; every additional concept; will have some nontrivial reduction in the number of people who make it through the learning curve gauntlet.

abrennan89 commented:
It doesn't look like there are any install instructions in the tutorial for `func`, it just goes straight in to https://deploy-preview-5243--knative.netlify.app/docs/getting-started/creating-function/? Is it going to be added as part of the quickstart plugin or something or is the install bit just yet to be added to docs?

abrennan89 commented:
Additionally, there should be a landing page IMO that explains the difference between functions vs services, and I think the functions section as a whole makes more sense _after_ the serving section. Reading https://deploy-preview-5243--knative.netlify.app/docs/getting-started/creating-function/ first it feels as if functions are the only way to create Knative apps, even though it doesn't explain Knative apps or services, and it's not clear any dependencies that are required before `func` can be used.

abrennan89 commented:
I don't think there's any need to include tabs either for the commands since there will only be one option (`kn func` or `func`, but not both)?

lance commented:
@abrennan89 @csantanapr I have added an installation section for `func`, opting for now to stick with just `func` as the CLI usage for the time being, pending further discussion.

I have also moved the functions section in the tutorial to the last, after serving and eventing. I think this probably should affect the copy in these sections somewhat, but if you think that serving and eventing concepts should be presented first, this is how it will look overall.

Personally, I think the benefit of functions is that users don't need to know anything about serving or eventing to get started with knative. To that point, if it were to remain at the top, perhaps it would be better to remove the whole section on using a broker and triggers, and instead just do an HTTP function.

Thank you for the feedback so far. I would love to have this published in the next couple of weeks, so please keep it coming if you have the opportunity.

lance commented:
Regarding `func` vs. `kn func`, another option would be for the `kn` CLI to include the `kn-plugin-func` in the client build by default. This would eliminate the need to provide `func` install instructions entirely, as it will already be available after having installed the `kn` CLI - for which installation is already documented.

salaboy commented:
This is great /lgtm

lkingland reviewed (commented):
This looks great!

My only suggestion below is that we downplay `func build` until later in the user's journey of discovery, focusing on `create` and `deploy`, and also in this same vein: this is the `func` docs, and should therefore focus on `func` only, with callouts for installing it as a plugin as a more advanced topic.

lkingland commented:
> I think the functions section as a whole makes more sense _after_ the serving section

This perhaps makes more sense for someone who already understands these topics, but teaching top-down yield better retention.  Start simple, elucidating more concepts as necessary.

lance commented on docs/getting-started/build-function.md:
```diff
@@ -0,0 +1,29 @@
+Before a function can be deployed, it must be built. The `func build` command
+creates an OCI container image that can be run locally on your computer or on
+a Kubernetes cluster with Knative. Knative functions use multiple strategies to
+build container images, including S2I and Cloud Native Buildpacks, both locally
```
I removed this file altogether and did not reference s2i or buildpacks in what I replaced it with.

salaboy commented:
@lance this an awesome tutorial!

csantanapr commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,56 @@
+In the `hello` directory, you will find the following files:
+
+```{ .console }
+tree hello
````
Need to provide to the user the command `cd hello` 

I think is better to tell the user `cd hello` and then `tree hello` take into account that mac user and windows user will not have `tree` in their terminal. Maybe add a note.

csantanapr commented on docs/getting-started/deploy-function.md:
````diff
@@ -0,0 +1,58 @@
+Deploy the function by running the command specifying the project directory:
+
+```{ .console }
+func deploy --registry <registry> -p hello
````
I would remove the extra flag `-p hello` and have the user work inside the `hello` directory.

abrennan89 commented:
1. I don't think there should be such a section as "More with Knative Functions" inside a tutorial / getting started section.

   The information included in https://deploy-preview-5243--knative.netlify.app/docs/getting-started/function-next-steps/ seems to be mostly conceptual explanations as opposed to step by step tasks for how a user gets from A to B, which is what a tutorial should be.

   This feels like it's supposed to be a workaround for not having a core Functions section in docs as we do for Serving and Eventing that can be pointed to for further reading, but I don't think that's an acceptable workaround IMO. I think there should be another issue opened to properly add a Functions docs section, not everything can or should live in the tutorial.

2. I'm still concerned about the structure here. In the previous discussions I thought it was decided that Functions would go after serving and eventing, but here, especially with the "More with Functions" section, it's both.

   Per earlier comments from Lance:
   > I have also moved the functions section in the tutorial to the last, after serving and eventing. I think this probably should affect the copy in these sections somewhat, but if you think that serving and eventing concepts should be presented first, this is how it will look overall.
   
   I'm not sure why now it's both before the serving and eventing sections and also after them. IMO it should still be after. Is this a decision that should be made by docs WG leads or am I missing something?

csantanapr commented on docs/getting-started/function-triggers.md:
````diff
@@ -0,0 +1,130 @@
+Create the function by running the command:
+
+```{ .console }
+func create -l go -t cloudevents echo
````
```suggestion
func create -l go -t cloudevents echo
cd echo
```
This avoids the user from needing to use `-p echo` in the next step, and if they are like me they can `code .` and work on IDE :-)

csantanapr commented on docs/getting-started/function-triggers.md:
````diff
@@ -0,0 +1,130 @@
+logs of the function pod. Watch the logs of the function pod by running the command:
+
+```{ .console }
+kubectl logs -l function.knative.dev=true --tail=10
````
Need to tell the user to wait, if they get this message to try again. I had to wait 15 seconds to get some output
```
➜  func-tutorial kubectl logs -l function.knative.dev=true --tail=10                                                

No resources found in default namespace.
```

evankanderson replied:
Any chance the function could provide a link to its own logs somehow on a GET? Probably not, but I thought it might be worth asking...

csantanapr commented on config/nav.yml:
```diff
@@ -31,6 +34,9 @@ nav:
           - Using a Knative Service as a source: getting-started/first-source.md
           - Using Triggers and sinks: getting-started/first-trigger.md
+        - More with Knative Functions:
+          - Next Steps: getting-started/function-next-steps.md
+          - Function Triggers and Event Sources: getting-started/function-triggers.md
```
```suggestion
         - Function Triggers and Event Sources: getting-started/function-triggers.md
          - Next Steps: getting-started/function-next-steps.md
```
I would put "Next Steps" last as there was nothing useful to run in my termal just docs and something about local builds that I feel too much work at this stage.

csantanapr commented on config/nav.yml:
```diff
@@ -31,6 +34,9 @@ nav:
           - Sources, Brokers, and Triggers: getting-started/first-broker.md
           - Using a Knative Service as a source: getting-started/first-source.md
           - Using Triggers and sinks: getting-started/first-trigger.md
+        - More with Knative Functions:
```
Why you have this split?
I think you should have the 4 pages under "Using Knative Functions" show the hello and echo back to back in one place.
On the navigation, if user clicks next it goes to Serving tutorial.

abrennan89 replied:
Yep, splitting this way to me is very confusing.

lance replied:
> Why you have this split?
> I think you should have the 4 pages under "Using Knative Functions" show the hello and echo back to back in one place.

This was meant as a compromise. I feel pretty strongly that functions create and deploy steps should come first, since they do not require any knative knowledge at all. Ashleigh did not like having some things like triggers introduced before the Knative Eventing section, so I split these out. Could we note in the section that discusses Triggers and such that the user might want to do that portion of the tutorial and then come back to Functions Next Steps? I don't think this way would be super clean either, but might be an improvement? WDYT?

nainaz commented:
Thank you @lance for providing the stop gap for the lack of Functions doc on knative/docs. It looks great. 
1. As a compromise until we have all the bits figured out around Documentation on Functions. Doing http first and second section after eventing explanation is okay. What we can say is that ready for Functions with eventing/cloud events. 
2. Since we need something related to functions in there and "kn func" is not default, we keep it func. but later we move to kn func. I agree keeping the experience consistent will go a long way. but as of now func is easier to install and start working with.

psschwei commented:
Installing `kn` is one of the steps in the "Install Knative using quickstart" section of the tutorial (which precedes the func ones), so users should already have it on their system before they get to the functions sections. Users will also already installed and used a plugin (quickstart) by this point as well.

lance commented:
> Installing `kn` is one of the steps in the "Install Knative using quickstart" section of the tutorial (which precedes the func ones), so users should already have it on their system before they get to the functions sections. Users will also already installed and used a plugin (quickstart) by this point as well.

I looked into how it might work to do this. If we really must use `kn func` for the tutorial, I would like to discuss how that looks in practice before implementing it.

* I would still like to keep this page: https://deploy-preview-5243--knative.netlify.app/docs/functions/install-func/. It provides instructions on installing the `func` CLI as standalone, and a link to the client documentation for installing plugins.
* Add an "Installing Knative Functions CLI" page to the Getting Started -> Using Knative Functions section, just before "Creating a Function". On this page have instructions for installing as a plugin, but note that you can install it standalone, linking to the installation page above.
* In the tutorial reimplement the tabs that were there with only `func`, but this time have two tabs, one with `func` and one with `kn func`.

Does this sound like a reasonable compromise?

lance commented:
@abrennan89 @csantanapr @salaboy @lkingland @psschwei thanks all of you for your input so far. I've reworked some things.

- Function installation has been removed from the quickstart install, since it's not required for the quickstart plugin per se
- A page on installing functions has been added as the first section in the Functions tutorial. It discusses how it can be used either standalone or as a plugin, and that if you already have `kn` installed via brew, installing `func` the same way will give you interchangable `kn func` and `func`.
- Tabs have been added to refer to both `func` and `kn func` where the function commands are used
- Text has been added that talks explicitly about the fact that developers do not need to know or understand knative primitives before using knative functions
- "Next with Knative Functions" has been removed from the navigation altogether and added as a link in the "What's next" page, and at the end of the "Function Event Sources" page.

I have *not* moved the functions tutorial below the Serving and Eventing sections, because I really don't think a user needs to walk through traffic splitting and autoscaling before they get to functions.

I'm asking for you to please take a look again. Thank you!

psschwei commented:
>  it's much easier to run `brew install func` than it is to have additional documentation around how to set up `func` as a plugin for `kn`

At least in the case of quickstart, doing `brew install quickstart` will put the binary in the proper place so that it's picked up by `kn` automatically, no other setup involved. 

> If we really must use kn func for the tutorial...

I think what's holding me up a bit here is that as the docs are written now we have two CLIs for core: `kn` for serving/eventing and `func` for functions. I'm assuming that's not what we want and that eventually we'd want to get the func CLI bits into `kn`. So using func as a plugin, i.e. `kn func`, just makes it so users don't have to relearn the pattern once that move is made.

On the other hand, if I'm assuming wrong and it's already been decided that `kn` and the `func` CLI bits are going to stay separate in the long run, then it's probably best to just use `func` in the tutorial. 

(and sorry if y'all have already gone over this and I'm rehashing old business...)

lance commented:
> I think what's holding me up a bit here is that as the docs are written now we have two CLIs for core: `kn` for serving/eventing and `func` for functions. I'm assuming that's not what we want and that eventually we'd want to get the func CLI bits into `kn`. So using func as a plugin, i.e. `kn func`, just makes it so users don't have to relearn the pattern once that move is made.

Hi @psschwei I have modified this PR so that both methods are shown using the tabbed options as are used elsewhere in the tutorial (e.g. "using kn" and "using yaml"). This allows users to use whatever method they are comfortable with. On the installation page, I have this copy (which I hope makes it very clear).

> The Knative Functions CLI (func) provides an interface for creating, configuring, building and deploying Knative Functions without the need to create or modify YAML files directly. The func CLI can be installed as a standalone binary or as a plugin to the Knative CLI (kn). If you already have kn installed via [Homebrew](https://brew.sh/), when you install the func CLI using [Homebrew](https://brew.sh/), it is automatically recognized as a plugin to kn and can be referenced as kn func or func interchangeably.

> On the other hand, if I'm assuming wrong and it's already been decided that kn and the func CLI bits are going to stay separate in the long run, then it's probably best to just use func in the tutorial.

When we moved the project to Knative, it was clear that the `func` CLI worked standalone. The `func` CLI is meant to be a different user experience than `kn` and cater to developers who do not know (or want to know) how everything works. To my knowledge there has been no directive to make `func` be __only__ a plugin for `kn` and eliminate the `func` standalone. If so, I would be quite surprised.

evankanderson commented on docs/getting-started/function-next-steps.md:
```diff
@@ -0,0 +1,58 @@
+# Next Steps with Knative Functions
```
I don't see this linked into the nav anywhere -- intended?

lance replied:
Currently it's only pointed to in the Next Steps document. I think ultimately this content (probably modified) should be in the main Functions section of the site once that has been created.

evankanderson commented on docs/getting-started/function-triggers.md:
```diff
@@ -0,0 +1,147 @@
+to CloudEvents that are sent by Knative Eventing resources.
+
+In this tutorial, you will create a simple Knative event source to generate
+CloudEvents, as well as other Knative Eventing resources. You don't need
+to know anything about Knative Eventing primitives to follow this tutorial.
+However, if you would like to learn more about Knative Eventing, see the
+[Using Knative Eventing](./getting-started-eventing.md) section of this tutorial.
```
```suggestion
CloudEvents, as well as other Knative Eventing resources. You'll learn more about Knative Eventing
[later on in this tutorial](./getting-started-eventing.md).
```

evankanderson commented on docs/getting-started/function-triggers.md:
```diff
@@ -0,0 +1,147 @@
+
+## Create a Broker
+
+Now, let's create the Broker. A Broker is a resource that will receive events
+from the event source and forward them to other resources.
+For many of the following commands, you will use the `kn` CLI. If you don't
+already have the `kn` CLI installed, please refer to the [Knative
+Client installation instructions](https://knative.dev/docs/client/install-kn/).
```
This should line up with the setup for the getting-started tutorial, which [requires having `kn` installed](https://knative.dev/docs/getting-started/).

evankanderson replied:
Also, less exposition -- maybe:

```suggestion
Next, we'll create a Broker to buffer events between the event source and the function.
```

evankanderson commented on docs/getting-started/function-triggers.md:
```diff
@@ -0,0 +1,147 @@
+already have the `kn` CLI installed, please refer to the [Knative
+Client installation instructions](https://knative.dev/docs/client/install-kn/).
+
+Create a default Broker by running the command:
```
```suggestion
Create a Broker named `default` by running the command:
```

evankanderson commented on docs/getting-started/function-triggers.md:
```diff
@@ -0,0 +1,147 @@
+
+## Create a PingSource
+
+A PingSource is a Knative Event Source that generates events at a specified interval.
+To create the PingSource, you will also use the `kn` CLI. This command will
+result in a PingSource emitting an event every minute, sending it to the
+default Broker.
+
+Create a PingSource called, `ping` by running the command:
```
```suggestion
To ensure we get events quickly, we'll create a [PingSource](https://knative.dev/docs/eventing/sources/ping-source/) which generates an event every minute. Note that the PingSource uses a `crontab`-type expression:
```

evankanderson commented on docs/getting-started/function-triggers.md:
````diff
@@ -0,0 +1,147 @@
+      knativearrivaltime: 2022-09-23T14:25:01.295357548Z
+    ```
+
+That's it! To learn
+more about local development with Knative Functions, see
+[Next Steps with Knative Functions](../function-next-steps).
+Or view the [Command Line Reference](https://github.com/knative-sandbox/kn-plugin-func/blob/main/docs/reference/func.md)
+for detailed descriptions of all CLI commands.
````
I think these should go in `/getting-started/next-steps.md`, not here (since this is in the middle of the tutorial flow, we don't want to explicitly route people out of the tutorial).

evankanderson commented on docs/getting-started/next-steps.md:
```diff
@@ -18,6 +18,7 @@ See the following guides for documentation specific to your use case:
 
 - [Serving Guide](../serving/README.md)
 - [Eventing Guide](../eventing/README.md)
+- [Next Steps With Functions](../function-next-steps/)
```
I think the `function-next-steps` doc should maybe be a `../functions/README.md`, like Serving and Eventing?

evankanderson reviewed (commented):
I think we want a top-level section on `Functions` after `Eventing`. (I'm okay if that's a "will do in another PR", but I want to make sure to capture that feedback.)

Thanks for writing this up!

lance commented on docs/snippets/install-func.md:
````diff
@@ -0,0 +1,84 @@
+
+    1. Build an executable binary:
+
+          ```bash
+          make
+          ```
````
Maybe `go install knative.dev/func@latest`? But when I try that I get 

```
❯ go install knative.dev/func@latest
go: downloading knative.dev/func v0.34.1
go: knative.dev/func@latest (in knative.dev/func@v0.34.1):
	The go.mod file for the module providing named packages contains one or
	more replace directives. It must not contain directives that would cause
	it to be interpreted differently than if it were the main module.
```

lkingland replied:
This should indeed work.
Referenced in a [new issue](https://github.com/knative/func/issues/1308)

lance commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,60 @@
+
+Create the function by running the command:
+
+=== "func"
+
+    ```{ .console}
+    func create hello
+    ```
+
+=== "kn func"
+
+    ```{ .console }
+    kn func create -l go hello
+    ```
````
I'm OK with that, but there has been a lot of back and forth over the use of `func` or `kn func` in the tutorial. This was an attempt to address these disagreements. Do you have thoughts on that?

lkingland replied:
In either case, should they both include the `-l go` flag?

lkingland replied:
Agree with Evan in the long run, but for now this seems like a reasonable compromise to get us past the acute hurdle.

lance commented:
> I think we want a top-level section on `Functions` after `Eventing`. (I'm okay if that's a "will do in another PR", but I want to make sure to capture that feedback.)

Opened an issue for this https://github.com/knative/docs/issues/5265

> Thanks for writing this up!

👍

lance commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,60 @@
+it to your cluster. Before you do that, let's take a look at the project.
+In the `hello` directory, you will find the following files:
+
+```{ .console }
+cd hello
+ls
+```
+
+!!! Success "Expected output"
+  ```{ .console .no-copy }
+   func.yaml   go.mod   handle.go   handle_test.go   README.md
+  ```
````
I think maybe that's good. But perhaps I should only show `handle.go` and maybe `handle_test.go`. The `func.yaml` file really is only necessary to know about for more advanced use cases. And `go.mod` and `README.md` should be self explanatory.

abrennan89 commented on docs/snippets/install-func.md:
````diff
@@ -0,0 +1,84 @@
+
+        Where `<path-to-binary-file>` is the path to the binary file you downloaded in the previous step, for example, `func_darwin_amd64` or `func_linux_amd64`.
+
+    1. Move the executable binary file to a directory on your PATH by running the command:
+
+        ```bash
+        mv func /usr/local/bin
+        ```
````
IMO another good reason to just have `kn func` as a plugin and leave `func` out...

lance replied:
Pretty much copied directly from the `kn` installation instructions. https://knative.dev/docs/client/install-kn/

abrennan89 commented on docs/getting-started/creating-function.md:
```diff
@@ -0,0 +1,60 @@
+# Creating a Knative Function
+
+Knative Functions provide a programming model that leverages the Knative
+Serving and Eventing APIs to provide a quick and easy way for developers
+to deploy their first Knative applications. You don't need to know anything
```
Again, this why why I think functions should be after serving and eventing...

lkingland replied:
>  I think functions should be after serving and eventing...

Bottom-up explanations are enticing for those doing the explaining, especailly when one has such broad and deep understanding already, but the words risk falling on deaf ears of eager learners because they have no context upon which to anchor their knowledge.

Instead starting from a context-free starting point, such as the task of deploying a new function which echoes your GET request, or prints "hello world", is easy to grasp.  From this beach-head of knowledge, gradually introducing more advanced topics (serving, eventing, traffic shaping, scaling, etc)  is much more likely to lead to long-term knowledge retention, and consequently adoption.

lkingland replied:
> * Allows you to gradually explore the rest of Knative and Kubernetes when you need it

^^^ this

abrennan89 commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,60 @@
+    Created go function in hello
+    ```
+
+That's it. You have created your first Knative Function. You can now deploy
+it to your cluster. Before you do that, let's take a look at the project.
+In the `hello` directory, you will find the following files:
````
I feel like
"That's it. You have created your first Knative Function. Let's take a look at the project before deploying it."
is very fluffy - can we make the text more minimal and remove some of the informal fluff?

lance commented on docs/getting-started/deploy-function.md:
````diff
@@ -0,0 +1,60 @@
+    Body:
+    ```
+
+Next, you will learn how to integrate a function with Knative Eventing.
````
I'm not really sure about that. It would require another build, which can be long-ish. What's the real value of doing it, given that the function body is so simple?

lkingland commented on docs/getting-started/creating-function.md:
````diff
@@ -0,0 +1,129 @@
+=== "func"
+
+    ```{ .console}
+    func create hello
````
Pretty sure this should include the `-l go` flag.

lkingland commented on docs/getting-started/deploy-function.md:
```diff
@@ -0,0 +1,70 @@
+If you followed the [Quickstart](./quickstart-install.md) section of the tutorial,
+you should already have a Docker daemon on your local machine.
+Deploying from a local build requires that you have this available.
+Additionally, you must have a Docker registry that you
```
```suggestion
Additionally, you must have a container registry that you
```

lance commented:
@abrennan89 @salaboy @csantanapr @lkingland @psschwei @evankanderson @snneji OK I have incorporated most feedback. I think this is probably ready to go. The only outstanding items are the contentious placement of the content, and the use of both `func` and `kn func`. I still feel that having Functions in the tutorial first makes sense, and I think having the tabs for both uses of the binary is a reasonable compromise. PTAL

zroubalik commented on docs/snippets/install-func.md:
````diff
@@ -0,0 +1,84 @@
+    1. Check out the `func` client repository:
+
+          ```bash
+          git clone https://github.com/knative-sandbox/kn-plugin-func.git func
````
```suggestion
          git clone https://github.com/knative/func.git
```

zroubalik commented on docs/snippets/install-func.md:
```diff
@@ -0,0 +1,84 @@
+
+    You can install `func` by downloading the executable binary for your system and placing it in the system path.
+
+    1. Download the binary for your system from the [`func` release page](https://github.com/knative-sandbox/kn-plugin-func/releases){target=_blank}.
```
```suggestion
    1. Download the binary for your system from the [`func` release page](https://github.com/knative/func/releases){target=_blank}.
```

zroubalik reviewed (commented):
This is great, awesome job Lance! Just a small nit, there are still some references (more than I highlighted here) to https://github.com/knative-sandbox/kn-plugin-func where it should have been https://github.com/knative/func

Let's merge it once this issue is resolved.

lance commented:
> This is great, awesome job Lance! Just a small nit, there are still some references (more than I highlighted here) to https://github.com/knative-sandbox/kn-plugin-func where it should have been https://github.com/knative/func
> 
> Let's merge it once this issue is resolved.

Ahh good catch! Fixing that now...

abrennan89 commented on config/nav.yml:
```diff
@@ -251,6 +257,13 @@ nav:
     # Eventing reference docs
       - Reference:
         - Eventing API: eventing/reference/eventing-api.md
+###############################################################################
+    # Functions
+###############################################################################
+    - Functions:
+      - About Knative Functions: functions/README.md
+      - Function Runtimes: functions/runtimes.md
+      - Local Development and Testing: functions/local-development.md
```
I feel like this should be a separate PR since we have another issue for it: https://github.com/knative/docs/issues/5265

lance replied:
Honestly, I would rather keep it all as one and we can follow up with updates if this is not sufficient. Any reason why one PR shouldn't close two issues?

abrennan89 commented on docs/getting-started/creating-function.md:
```diff
@@ -0,0 +1,129 @@
+# Creating a Knative Function
+
+Functions are the easiest path to get started with
+Knative, providing a programming model that leverages the Knative
+Serving and Eventing APIs for a quick and easy method of deploying your
+first Knative applications. You don't need to know anything
+about Knative or even Kubernetes resources to get started.
```
Suggest removing, fluff

abrennan89 reviewed (changes requested):
Recommend removing all the fluff text and replacing with a simple explanation of what functions are and why a user wants to use them.

Also, please make all headings sentence case, and use all general guidelines from the docs style guide: https://github.com/knative/docs/tree/main/contribute-to-docs/style-guide

trisberg commented on docs/getting-started/function-triggers.md:
````diff
@@ -0,0 +1,151 @@
+Create a Trigger named `ping-trigger` by running the command:
+
+```{ .console }
+kn trigger create ping-trigger --sink hello
````
I think this should use the `echo` function created a bit earlier

lance commented:
@abrennan89 thanks for the feedback! 👍  I've incorporated most of your comments, but I do still have two outstanding questions/issues.

- If there is no real reason not to, why not just include the `/functions/*` stuff in this PR to close out #5265? Then we can iterate on that section by creating issues specific to each page/section/idea. 
- I'm open to suggestion for content in `/functions/README.md`. I think that what's there could be a reasonable start with an eye towards longer-term refinement, but if you have specific content that you would like to see there - or modifications to what I've written, that would be welcome.

psschwei commented:
> @abrennan89 @salaboy @csantanapr @lkingland @psschwei @evankanderson @snneji OK I have incorporated most feedback. I think this is probably ready to go. The only outstanding items are the contentious placement of the content, and the use of both `func` and `kn func`. I still feel that having Functions in the tutorial first makes sense, and I think having the tabs for both uses of the binary is a reasonable compromise. PTAL

We discussed `func` vs. `kn func` on the TOC call today, with the plan being to leave as is for the upcoming release (1.8) and will look into possibly including func in the client build for the release after that (1.9). So it's a LGTM for me.

csantanapr commented:
Thanks @lance for incorporating the Functions top level section. this lgtm

abrennan89 commented:
Myself and @lance have discussed offline providing docs WG help for this. Please hold off on merging this PR for the moment, since there is additional work happening from my end on Functions docs and we plan to sync on this next week.

/hold

abrennan89 commented:
Closing this PR since the work is being done elsewhere - see https://github.com/knative/docs/issues/5265 for tracking
